### ISSUE TITLE
Saga common for all routes

### ISSUE THREAD

**Issue by williamgranli** [mxstbr/react-boilerplate#1137]
I have a component called Auth which has sagas for logging in/out and registering. The logout function should be listening on all routes (except login) so that the user can log out on any page. 

Right now I'm importing the logoutFlow saga from the Auth component in all my other saga files which doesn't seem like the best solution. 

Is there a nicer way to activate this saga for all routes?

**Comment by sedubois** [comment]
You could simply import the saga from the `App` component. Can also take inspiration from this pull request https://github.com/mxstbr/react-boilerplate/pull/1108. (There, so far there's no saga for logout, just for the login callback.)

**Comment by williamgranli** [comment]
@sedubois sorry, but I'm new to react and redux... how do I import and activate the saga from the App component? I'm struggling since it's not a route, so it can't be done the same way as in routes.js. 

I can't do this either (according to saga examples), since the store has already been created: 
`sagaMiddleware.run(logoutFlow)`

Any help is much appreciated!

**Comment by sedubois** [comment]
@williamgranli you can actually import your ad-hoc saga from `routes.js`, and add `injectSagas(saga)` at the beginning of `createRoutes`. E.g:

```
import ActionWatcher from 'containers/ActionWatcher/sagas';
...
export default function createRoutes(store) {
  const { injectReducer, injectSagas } = getAsyncInjectors(store);
  injectSagas(ActionWatcher);
  return [...];
}
```

```
/* eslint-disable */

import { take } from 'redux-saga/effects';

export function* actionWatcher() {
  let action;
  while (action = yield take()) {
    console.log('got action!', action);
  }
}

// Bootstrap sagas
export default [
  actionWatcher,
];
```

**Comment by carlos-peru** [comment]
@williamgranli, how did you solve the logout saga issue? I was trying to do the same. I did what @sedubois suggested (thanks!) and it did kick the saga in the children routes although when redux-auth-wrapper redirects to /login (when accessing a protected route) the saga gets cancelled and I don't know why.

However, now I'm thinking that the logout flow should be synchronous and just have a loading page in between since it's just a local update (set state to null and/or set to null localstorage value). Any input would be appreciated.

**Comment by williamgranli** [comment]
@carlos-peru Not sure what you are asking exactly, but I did what sedbois suggested and it works fine for me. 

My routes.js:
```

import { logoutFlow, loginFlow } from './containers/Authentication/sagas';

...

export default function createRoutes(store) {
  // Create reusable async injectors using getAsyncInjectors factory
  const { injectReducer, injectSagas } = getAsyncInjectors(store); // eslint-disable-line no-unused-vars

  // Inject logoutFlow to make it available to all routes
  injectSagas([loginFlow, logoutFlow]);

...
```

Could you try to specify what is causing your issue? (Steps on how to reproduce if possible).

**Comment by carlos-peru** [comment]
Well I did that, but when the redux-auth-wrapper would redirect me from let's say /dashboard to /login for validation, the saga would get cancelled. However, it would work when I'd go straight to /login. I don't know why. I ended up moving back the saga to the LoginPage container but kept the reducer, selector and constants in App. My LogoutLink container also uses the same reducer, selector and constants. Many of the answers you got in different posts helped put together. Thanks @williamgranli!

**Comment by zsherman** [comment]
@carlos-peru that also happened to me :/

**Comment by zsherman** [comment]
Actually @carlos-peru realized that if you're using the boilerplate this is suggested:

```
yield take(LOCATION_CHANGE);
yield cancel(watcher);
```

which will cancel the logout generator after your login redirect.

**Comment by williamgranli** [comment]
Doing what @zsherman said as well, as below:

```
export function* root() {
  const watcher1 = yield fork(saveIntegrationFlow);
  const watcher2 = yield fork(getConnectorsFlow);

  yield take(LOCATION_CHANGE);
  yield cancel(watcher1);
  yield cancel(watcher2);
}

// All sagas to be loaded
export default [
  root,
];
```

**Comment by justingreenberg** [comment]
this should be resolved by #1746 — feedback is welcome!
